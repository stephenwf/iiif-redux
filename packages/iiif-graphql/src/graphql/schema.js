import { makeExecutableSchema } from 'graphql-tools';

import typeDefs from '../schema.graphql';
import resolvers from '../resolvers';
import schemaDirectives from './directives';

const schema = makeExecutableSchema({
  typeDefs,
  resolvers,
  schemaDirectives,
});

export default schema;
