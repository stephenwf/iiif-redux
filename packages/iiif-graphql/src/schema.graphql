directive @intl on FIELD_DEFINITION

interface CollectionItem {
  # Technical
  id: String
  type: String
  viewingDirection: String
  behaviour: String

  # Descriptive
  label: String @intl
  metadata: [Metadata]
  summary: String @intl
  thumbnail: [ContentResource]
  posterCanvas: Canvas
  requiredStatement: String @intl
  rights: String
  navDate: String

  # Linking
  seeAlso: [ContentResource]
  service: [Service]
  logo: [ContentResource]
  homepage: ContentResource
  rendering: [ContentResource]
  partOf: PartOf
}

type Collection implements CollectionItem {
  # Technical
  id: String
  type: String
  viewingDirection: String
  behaviour: String

  # Descriptive
  label: String @intl
  metadata: [Metadata]
  summary: String @intl
  thumbnail: [ContentResource]
  posterCanvas: Canvas
  requiredStatement: String @intl
  rights: String
  navDate: String

  # Linking
  seeAlso: [ContentResource]
  service: [Service]
  logo: [ContentResource]
  homepage: ContentResource
  rendering: [ContentResource]
  partOf: PartOf

  # Structural
  items: [CollectionItem]
  annotations: [AnnotationPage]
}

type Manifest implements CollectionItem {
  # Technical
  id: String
  type: String
  viewingDirection: String
  behaviour: String

  # Descriptive
  label: String @intl
  metadata: [Metadata]
  summary: String @intl
  thumbnail: [ContentResource]
  posterCanvas: Canvas
  requiredStatement: String @intl
  rights: String
  navDate: String

  # Linking
  seeAlso: [ContentResource]
  service: [Service]
  logo: [ContentResource]
  homepage: ContentResource
  rendering: [ContentResource]
  partOf: PartOf

  # Structural
  items: [Canvas]
  canvases: [Canvas]
  annotations: [AnnotationPage]
  structures: [Range]
}

type Canvas {
  # Technical
  id: String
  type: String
  viewingDirection: String
  behaviour: String
  height: Int
  width: Int
  duration: Int

  # Descriptive
  label: String @intl
  metadata: [Metadata]
  summary: String @intl
  thumbnail: [ContentResource]
  posterCanvas: Canvas
  requiredStatement: String @intl
  rights: String
  navDate: String

  # Linking
  seeAlso: [ContentResource]
  service: [Service]
  logo: [ContentResource]
  homepage: ContentResource
  rendering: [ContentResource]
  partOf: PartOf

  # Structural
  items: [AnnotationPage]
  paintingAnnotations: [AnnotationPage]
  annotations: [AnnotationPage]
}

type Annotation {
  # Technical
  id: String
  type: String
  viewingDirection: String
  behaviour: String
  timeMode: String

  # Descriptive
  label: String @intl
  metadata: [Metadata]
  summary: String @intl
  thumbnail: [ContentResource]
  requiredStatement: String @intl
  rights: String

  # Linking
  seeAlso: [ContentResource]
  service: [Service]
  logo: [ContentResource]
  homepage: ContentResource
  rendering: [ContentResource]
  partOf: PartOf

  # Structural
  body: [AnnotationBody]
  target: [Selector]
}

type AnnotationCollection {
  # Technical
  id: String
  type: String
  viewingDirection: String
  behaviour: String
  timeMode: String

  # Descriptive
  label: String @intl
  metadata: [Metadata]
  summary: String @intl
  thumbnail: [ContentResource]
  requiredStatement: String @intl
  rights: String

  # Linking
  seeAlso: [ContentResource]
  service: [Service]
  logo: [ContentResource]
  homepage: ContentResource
  rendering: [ContentResource]
  partOf: PartOf

  # Structural
  items: [AnnotationPage]
  first: AnnotationPage
  last: AnnotationPage
}

type AnnotationPage {
  # Technical
  id: String
  type: String
  viewingDirection: String
  behaviour: String
  timeMode: String

  # Descriptive
  label: String @intl
  metadata: [Metadata]
  summary: String @intl
  thumbnail: [ContentResource]
  requiredStatement: String @intl
  rights: String

  # Linking
  seeAlso: [ContentResource]
  service: [Service]
  logo: [ContentResource]
  homepage: ContentResource
  rendering: [ContentResource]
  partOf: PartOf

  # Structural
  next: AnnotationPage
  previous: AnnotationPage
  items: [Annotation]
  annotations: [Annotation]
}

type Range {
  # Technical
  id: String
  type: String
  viewingDirection: String
  behaviour: String

  # Descriptive
  label: String @intl
  metadata: [Metadata]
  summary: String @intl
  thumbnail: [ContentResource]
  posterCanvas: Canvas
  requiredStatement: String @intl
  rights: String
  navDate: String

  # Linking
  seeAlso: [ContentResource]
  service: [Service]
  logo: [ContentResource]
  homepage: ContentResource
  rendering: [ContentResource]
  partOf: PartOf
  start: Canvas
  supplementary: AnnotationCollection

  # Structural
  items: [RangeItem]
  annotations: [AnnotationPage]
}

type ContentResource {
  id: String
}

union RangeItem = Range | Canvas

# Maybe union?
type CanvasOrReference {
  id: String
}

type Selector {
  id: String
}

type Metadata {
  label: String @intl
  value: String @intl
}

type AnnotationBody {
  value: String
}

type PartOf {
  id: String
}

type Service {
  id: String
}

type Choice {
  id: String
}

type Query {
  collection(id: String): Collection
  manifest(id: String): Manifest
}
