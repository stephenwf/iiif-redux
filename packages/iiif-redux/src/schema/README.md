# Presentation 2.x and 3.x

This library aims to give full coverage for both presentation 2 and presentation 3 APIs both for importing and selecting content.

When you import Presentation 2 or 3 content it will automatically be detected and imported. Mixed Presentation 2 and 3 content in a single document is
not supported. When importing presentation 2 content, due to inconsistencies you may find the original documents have been modified. This is intentional
to provide a predictable and normalized structure. Descriptive properties also get expanded out to the international specification, defaulting to english,
for these properties. Future exporting tools may revert this during an export. Future exporting will also offer an option to upgrade presentation 2 documents
to the latest stable presentation version.

## API Changes

When presentation updates happen, some parts of the API are deprecated and finally removed, and as such this library does not support calling deprecated
APIs on sources that are not the latest version. For example, between Presentation 2.1 and 3.0 the property `viewingHint` was changed to `behavior`. Along
with this change the scope of the property changed to expand what `behavior` may contain. Because of this, the following behavior applies:

- Calling `getViewingHint` on Presentation 2 document will return `document.viewingHint` (deprecation notice logged in development)
- Calling `getBehavior` on Presentation 2 document will return `document.viewingHint`
- Calling `getViewingHint` on Presentation 3 document will return `null` (deprecation notice logged in development)
- Calling `getBehavior` on Presentation 3 document will return `document.behavior`

Although this library does aim to have full coverage of all presentation versions, it also provides an upgrade path where possible from lower versions of
the specification in your application. If remove all deprecation notices from your application, your application will be compatible with the latest versions
of the specification.

Note: This compatibility only extends to structural differences, and not semantics that may have changed.

## Opting for strict selectors

If you know all of your content conforms strictly to one version of the specification, you can access the selectors directly:

```js
import { canvasByIdSelector } from 'iiif-redux/api/2.x/canvas';
// or
import { canvasByIdSelector } from 'iiif-redux/api/3.x/canvas';

// instead of
import { canvasByIdSelector } from 'iiif-redux/api/canvas';
```

Similarly, if you want to access the IIIF specification compliant selectors (maybe for your own library) you can access them here:

```js
import * as descriptive from 'iiif-redux/api/2.x/iiif/descriptive';
```

Its possible choosing a strict version implementation may improve performance on large documents. This will have undesired effects on the wrong
presentation versions though, so it should only be used on data-sets you control and not driven by user-input like traditional viewers.

## Sequences between 2.x and 3.x

Sequences are a special case in the change between versions 2 and 3. In Presentation 3 you can still call `getSequences` and `getDefaultSequence` on a manifest, but
instead of this being defined in the structure, its a special range that's returned. This acts in the same way as the sequence in presentation 2 and can be used to
reflect the same structure. However, its more common to instead just use the canvases directly off of the manifest. Because of this, presentation 2 has been retro-fitted
with a `getCanvases` selector that will skip the sequence. This is common practice for viewers.

## Language properties

In Presentation 2 and 3 descriptive (human-readable) strings are stored native to their presentation version. When you select, say, a label, you get returned one of 2 structures
representing the string:

```js
// Presentation 2
const p2Label = [{ '@language': 'en', '@value': ['some label'] }];

// Presentation 3
const p3Label = {
  en: ['some label'],
};
```

Its up to the end implementation to support these 2 language implementations. IIIF Redux does offer a function for rendering these strings.

```js
import { intlString } from 'iiif-redux/utility';

const englishString = intlString('en');

// Both return 'some label'
englishString([{ '@language': 'en', '@value': ['some label'] }]);

englishString({
  en: ['some label'],
});
```

## Annotation changes

Annotations have changed a lot between Presentation 2 and 3. It is still unknown how IIIF Redux will differ between implementations. There is a
1-to-1 parity between structural components, which should help to align the changes.

## Paging

Annotation paging defined in W3C remains, but is not supported in Presentation 2 or 3 in this library due to its relatively low usage. This may change
if demand changes.

## Implementation details

When you import a document either manually specifying a presentation version, or if you choose it to be auto-detected, for every resource imported an
index will be created that contains the schema version.

```js
const state = {
  schemaVersions: {
    'https://../1': 3,
    'https://../2': 3,
    'https://../3': 2,
  },
};
```

Although this does add an overhead to the redux store, it ensures that any resource that is loaded is identified correctly when you apply a selector.

You can opt-out of this extra overhead by creating a store targeting a specific version

## Presentation 3 questions

- For the field `seeAlso` and `partOf`, the specification does not provide a list of valid IIIF resources that can be found here, implying all IIIF resources are valid.
- Annotation pages only have one required field `type` which seems counter productive as it cannot function as an annotation page without annotations, should ID also be required?

## Changes to selector framework

Previously if you wanted to get just the label selector, you could do something like this:

```js
import collection from 'iiif-redux/api/collection';

const state = {
  /* ... */
};

const getFirstCollection = () => /***/ null;

const { getLabel } = collection(getFirstCollection);

const label = getLabel(state);

console.log(label);
```

So you could effectively have access to the individual selectors. Because of the uncertainty using multiple presentation versions, this model has now changed to match
the select by ID and collection selectors.

```js
import collection from 'iiif-redux/api/collection';

const state = {
  /* ... */
};

const getFirstCollection = () => /***/ null;

const firstCollection = collection(getFirstCollection);

const selector = firstCollection(api => ({
  label: api.getLabel,
}));

console.log(selector(state).label);

// Alternatively
const getLabel = firstCollection(api => api.getLabel);

const label = getLabel(state);

console.log(label);
```

This unifies how the other selectors work too. If you still want access to individual selectors, say to build up your own library, you can still access each selector
library for each presentation version:

```js
import presentation2Collection from 'iiif-redux/api/2.x/collection';
import presentation3Collection from 'iiif-redux/api/3.x/collection';

const getFirstCollection = () => /***/ null;

const { getLabel, getId } = presentation2Collection(getFirstCollection);

const myCustomP2Selector = createStructuredSelector({
  label: getLabel,
  id: getId,
});
```

This is only recommended if you know what version of the specification you are using as you will have to manage the different versions yourself at this point.

If there is demand for a universal API that exposes single selectors, it may be added and work like this:

```js
import collection from 'iiif-redux/api/2.x-3.x/collection';

const getFirstCollection = () => /***/ null;

const { getLabel, getId } = collection(getFirstCollection);
```

But this is likely to have a negative impact on performance as the version look up will happen for each individual field, instead of once per selector.
