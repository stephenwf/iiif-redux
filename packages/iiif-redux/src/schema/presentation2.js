/**
 * Presentation 2 - schema definition
 * ============================================================================
 *
 * - Collection
 * - Manifest
 * - Sequence
 * - Canvas
 * - Annotation
 * - AnnotationList
 * - Range
 * - Layer
 * - Image Content (Annotation)
 * - Other Content (Annotation List)
 *
 * - External resources
 * - Services
 */

// @todo Image Service API.
import { compose } from 'redux';
import { schema, normalize } from 'normalizr';
import moveStartCanvasToSequence from '../compat/moveStartCanvasToSequence';
import preprocessLinkedEntities from '../compat/preprocessLinkedEntities';
import addMissingIds from '../compat/addMissingIds';

function createEntity(name) {
  const options = {
    idAttribute: '@id',
  };

  return new schema.Entity(name, {}, options);
}

const collection = createEntity('collections');
const manifest = createEntity('manifests');
const sequence = createEntity('sequences');
const canvas = createEntity('canvases');
const annotation = createEntity('annotations');
const annotationList = createEntity('annotationLists');
const range = createEntity('ranges');
const layer = createEntity('layers');
const imageResource = createEntity('imageResources'); // thumbnails + image service
const choice = createEntity('choices');

// Unofficial types?
const externalResource = createEntity('externalResources');
const service = createEntity('services', true);
const within = new schema.Array(
  {
    collection,
    manifest,
    externalResource,
    layer,
  },
  entity => {
    switch (entity['@type']) {
      case 'sc:Collection':
        return 'collection';
      case 'sc:Manifest':
        return 'manifest';
      case 'sc:Layer':
        return 'layer';
      default:
        return 'externalResource';
    }
  }
);

const rangeMember = new schema.Union(
  {
    canvas,
    range,
  },
  input => (input['@type'] === 'sc:Canvas' ? 'canvas' : 'range')
);

/**
 * Dereferencable resources:
 *
 * - Collection (Required)
 * - Manifest (Required)
 * - First Sequence (Optional)
 * - Nth Sequence (Required)
 * - Canvas (Optional)
 * - Annotation (Recommended)
 * - AnnotationList (Required)
 * - Range (Optional)
 * - Layer (Optional)
 * - Image Content (Required)
 * - Other Content (Required)
 */
const RESOURCE_TYPE_MAP = {
  'sc:Collection': 'collection',
  'sc:Sequence': 'sequence',
  'sc:Manifest': 'manifest',
  'sc:Canvas': 'canvas',
  'sc:AnnotationList': 'annotationList',
  'sc:Annotation': 'annotation',
  'sc:Range': 'range',
  'sc:Layer': 'layer',
  'dctypes:Image': 'imageResource',
};
const resource = new schema.Union(
  {
    collection,
    sequence,
    manifest,
    canvas,
    annotationList,
    annotation,
    range,
    layer,
    imageResource,
    service,
  },
  entity => {
    if (RESOURCE_TYPE_MAP[entity['@type']]) {
      return RESOURCE_TYPE_MAP[entity['@type']];
    }
    if (entity.profile) {
      return 'service';
    }
    throw Error('Unknown entity type');
  }
);

// ===========================================================================
// 1) Collections
// ===========================================================================
collection.define({
  collections: new schema.Array(
    {
      collection,
      manifest,
    },
    entity => (entity['@type'] === 'sc:Manifest' ? 'manifest' : 'collection')
  ),
  manifests: [manifest],
  members: new schema.Array(
    {
      collection,
      manifest,
    },
    entity => (entity['@type'] === 'sc:Manifest' ? 'manifest' : 'collection')
  ),

  // Linking
  seeAlso: [externalResource],
  service: [service],
  related: [externalResource],
  rendering: [externalResource],
  otherContent: [annotationList],
  within: within,

  // Extra
  thumbnail: imageResource,
});

// ===========================================================================
// 2) Manifests
// ===========================================================================
manifest.define({
  sequences: [sequence],
  structures: [range],

  // Linking
  seeAlso: [externalResource],
  service: [service],
  related: [externalResource],
  rendering: [externalResource],
  otherContent: [annotationList],
  within: within,

  // Extra
  thumbnail: imageResource,
});

// ===========================================================================
// 3) Sequences
// ===========================================================================
sequence.define({
  canvases: [canvas],

  // Linking
  seeAlso: [externalResource],
  service: [service],
  related: [externalResource],
  rendering: [externalResource],
  otherContent: [annotationList],
  within: within,
  startCanvas: canvas,

  // Extra
  thumbnail: imageResource,
});

// ===========================================================================
// 4) Canvases
// ===========================================================================
canvas.define({
  images: [annotation],

  // Linking
  seeAlso: [externalResource],
  service: [service],
  related: [externalResource],
  rendering: [externalResource],
  otherContent: [annotationList],
  within: within,

  // Extra
  thumbnail: imageResource,
});

// ===========================================================================
// 5) Annotation
// ===========================================================================
annotation.define({
  resource: new schema.Union(
    {
      imageResource,
      externalResource,
      choice,
    },
    input => {
      switch (input['@type']) {
        case 'dctypes:Image':
          return 'imageResource';
        case 'oa:Choice':
          return 'choice';
        default:
          return 'externalResource';
      }
    }
  ),
  on: canvas,

  // Linking
  seeAlso: [externalResource],
  service: [service],
  related: [externalResource],
  rendering: [externalResource],
  within: within,

  // Extra
  thumbnail: imageResource,
});

// ===========================================================================
// 6) Annotation List
// ===========================================================================
annotationList.define({
  resources: [annotation],

  // Linking
  seeAlso: [externalResource],
  service: [service],
  related: [externalResource],
  rendering: [externalResource],
  within: within,

  // Extra
  thumbnail: imageResource,
});

// ===========================================================================
// 7) Range
// ===========================================================================
range.define({
  resources: [annotation],
  members: [rangeMember],
  canvases: [canvas],
  ranges: [range],

  // Linking
  seeAlso: [externalResource],
  service: [service],
  related: [externalResource],
  rendering: [externalResource],
  within: within,
  startCanvas: canvas,
  contentLayer: layer,

  // Extra
  thumbnail: imageResource,
});

// ===========================================================================
// 8) Layer
// ===========================================================================
layer.define({
  // Linking
  seeAlso: [externalResource],
  service: [service],
  related: [externalResource],
  rendering: [externalResource],
  within: within,
});

// ===========================================================================
// 9) Image resource
// ===========================================================================
imageResource.define({
  service: [service],
});

// ===========================================================================
// 10) Choice
// ===========================================================================
choice.define({
  default: annotation,
  item: [annotation],
});

const preprocess = compose(
  moveStartCanvasToSequence,
  preprocessLinkedEntities,
  addMissingIds
);

const normalizeResource = (rawResource, customSchema = resource) =>
  normalize(preprocess(rawResource), customSchema);

export {
  normalizeResource as normalize,
  preprocess,
  resource,
  collection,
  manifest,
  sequence,
  canvas,
  annotation,
  annotationList,
  range,
  layer,
  imageResource,
  externalResource,
  within,
  service,
};
