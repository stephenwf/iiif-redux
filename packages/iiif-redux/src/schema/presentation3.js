import { schema, normalize } from 'normalizr';
import { compose } from 'redux';
import addMissingIds from '../compat/addMissingIds';
import parseSelectorTarget from '../utility/parseTargetSelector';

function createEntity(name, processStrategy) {
  const options = {
    idAttribute: 'id',
  };
  if (processStrategy) {
    options.processStrategy = processStrategy;
  }

  return new schema.Entity(name, {}, options);
}

const collection = createEntity('collections'); // 1
const manifest = createEntity('manifests'); // 2
const canvas = createEntity('canvases'); // 3
const annotation = createEntity('annotations', original => {
  const body = Array.isArray(original.body) ? original.body : [original.body];
  const target = Array.isArray(original.target)
    ? original.target
    : [original.target];
  const newObj = { ...original };
  if (original.body) {
    newObj.body = body;
  }
  if (original.target) {
    newObj.target = target;
  }
  return newObj;
}); // 4
const annotationPage = createEntity('annotationPages'); // 5
const annotationCollection = createEntity('annotationCollections'); // 6
const range = createEntity('ranges'); // 7
const contentResource = createEntity('contentResources'); // 8
const choice = createEntity('choices'); // 9
const canvasReference = createEntity('canvasReferences'); // 10
const selector = new schema.Entity(
  'selectors',
  {
    id: canvas,
  },
  {
    processStrategy(value, parent, key) {
      // @todo This is an impossible path to hit. The process strategy is a strange function that
      // doesn't always hit the containing code. Will need to review at a later date.
      // if (typeof value === 'string') {
      //   return parseSelectorTarget(value);
      // }
      const id = value['@id'] || value.id;
      if (id) {
        return parseSelectorTarget(id);
      }
      return value;
    },
  }
);

// Unofficial types.
const service = createEntity('services');

// Union types
const partOf = new schema.Array(
  {
    collection,
    manifest,
    contentResource,
    annotationCollection,
  },
  entity => {
    switch (entity.type) {
      case 'Collection':
        return 'collection';
      case 'Manifest':
        return 'manifest';
      case 'AnnotationCollection':
        return 'annotationCollection';
      default:
        return 'contentResource';
    }
  }
);

const canvasOrReference = new schema.Union(
  {
    selector,
    canvasReference,
    canvas,
  },
  entity => {
    if (entity && entity.id && entity.id.indexOf('#') !== -1) {
      return 'selector';
    }
    if (entity.type === 'SpecificResource') {
      return 'canvasReference';
    }
    return 'canvas';
  }
);

const annotationBodyMappings = {
  Choice: 'choice',
  Application: 'contentResource',
  Dataset: 'contentResource',
  Image: 'contentResource',
  Sound: 'contentResource',
  Text: 'contentResource',
  Video: 'contentResource',
  TextualBody: 'contentResource',
};
const annotationBody = new schema.Union(
  {
    contentResource,
    choice,
  },
  input => {
    if (annotationBodyMappings[input.type]) {
      return annotationBodyMappings[input.type];
    }
    return 'contentResource';
  }
);

const rangeItem = new schema.Union(
  {
    canvas,
    selector,
    range,
    canvasReference,
  },
  input => {
    if (input && input.id && input.id.indexOf('#') !== -1) {
      return 'selector';
    }

    switch (input.type) {
      case 'Canvas':
        return 'canvas';
      case 'Range':
        return 'range';
      case 'SpecificResource':
      default:
        // @todo default?
        return 'canvasReference';
    }
  }
);

// Root resource
/**
 * Dereferencable resources:
 *
 * - Collection (Required)
 * - Manifest (Required)
 * - Canvas (Recommended)
 * - Annotation (Recommended)
 * - AnnotationPage (Required)
 * - AnnotationCollection (Optional)
 * - Content Resource (Required)
 * - Range (Optional)
 */
const RESOURCE_TYPE_MAP = {
  Collection: 'collection',
  Sequence: 'sequence',
  Manifest: 'manifest',
  Canvas: 'canvas',
  AnnotationCollection: 'annotationCollection',
  AnnotationPage: 'annotationPage',
  Annotation: 'annotation',
  Range: 'range',
  // Content resources.
  Application: 'contentResource',
  Dataset: 'contentResource',
  Image: 'contentResource',
  Sound: 'contentResource',
  Text: 'contentResource',
  Video: 'contentResource',
  TextualBody: 'contentResource',
};
const resource = new schema.Union(
  {
    collection,
    manifest,
    canvas,
    annotationCollection,
    annotationPage,
    annotation,
    range,
    contentResource,
    service,
  },
  entity => {
    if (RESOURCE_TYPE_MAP[entity.type]) {
      return RESOURCE_TYPE_MAP[entity.type];
    }
    if (entity.profile) {
      return 'service';
    }
    throw Error('Unknown entity type');
  }
);

// ===========================================================================
// 1) Collections
// ===========================================================================
collection.define({
  // Structural
  items: new schema.Array(
    {
      collection,
      manifest,
    },
    entity => (entity.type === 'Manifest' ? 'manifest' : 'collection')
  ),
  annotations: [annotationPage],

  // Linking.
  seeAlso: [contentResource],
  service: [service],
  logo: [contentResource],
  homepage: contentResource,
  rendering: [contentResource],
  partOf: partOf,

  // Extra.
  thumbnail: [contentResource],
  posterCanvas: canvas,
});

// ===========================================================================
// 2) Manifests
// ===========================================================================
manifest.define({
  items: [canvas],
  structures: [range],
  annotations: [annotationPage],

  // Linking
  seeAlso: [contentResource],
  service: [service],
  logo: [contentResource],
  homepage: contentResource,
  rendering: [contentResource],
  partOf: partOf,
  start: canvasOrReference,

  // Extra
  thumbnail: [contentResource],
  posterCanvas: canvas,
});

// ===========================================================================
// 3) Canvases
// ===========================================================================
canvas.define({
  // Structural.
  items: [annotationPage],
  annotations: [annotationPage],

  // Linking
  seeAlso: [contentResource],
  service: [service],
  logo: [contentResource],
  homepage: contentResource,
  rendering: [contentResource],
  partOf: partOf,

  // Extra
  thumbnail: [contentResource],
  posterCanvas: canvas,
});

// ===========================================================================
// 4) Annotation
// ===========================================================================
annotation.define({
  // Structural.
  body: [annotationBody],
  target: [selector],

  // Linking
  seeAlso: [contentResource],
  service: [service],
  logo: [contentResource],
  homepage: contentResource,
  rendering: [contentResource],
  partOf: partOf,

  // Extra
  thumbnail: [contentResource],
});

// ===========================================================================
// 5) Annotation Collection
// ===========================================================================
annotationCollection.define({
  // Structural.
  items: [annotationPage],

  // Linking
  seeAlso: [contentResource],
  service: [service],
  logo: [contentResource],
  homepage: contentResource,
  rendering: [contentResource],
  partOf: partOf,

  // extra
  thumbnail: [contentResource],

  // Paging.
  first: annotationPage,
  last: annotationPage,
});

// ===========================================================================
// 6) Annotation Page
// ===========================================================================
annotationPage.define({
  // Structural.
  items: [annotation],

  // Linking
  seeAlso: [contentResource],
  service: [service],
  logo: [contentResource],
  homepage: contentResource,
  rendering: [contentResource],
  partOf: partOf,

  // Extra
  thumbnail: [contentResource],

  // Paging.
  next: annotationPage,
  previous: annotationPage,
});

// ===========================================================================
// 7) Range
// ===========================================================================
range.define({
  // Structural.
  items: [rangeItem],
  annotations: [annotationPage],

  // Linking
  seeAlso: [contentResource],
  service: [service],
  logo: [contentResource],
  homepage: contentResource,
  rendering: [contentResource],
  partOf: partOf,
  start: canvasOrReference,
  supplementary: annotationCollection,

  // Extra
  thumbnail: [contentResource],
});

// ===========================================================================
// 8) Content Resource
// ===========================================================================
contentResource.define({
  // Linking.
  seeAlso: [contentResource],
  service: [service],
  logo: [contentResource],
  rendering: [contentResource],
  partOf: partOf,
  // Structural
  annotations: [annotationPage],
});

// ===========================================================================
// 9) Choice
// ===========================================================================
choice.define({
  default: annotation,
  item: [annotation],
});

// ===========================================================================
// 10) Canvas Reference
// ===========================================================================
canvasReference.define({
  source: canvas,
});

const preprocess = compose(addMissingIds(3.0));

const normalizeResource = (rawResource, customSchema = resource) =>
  normalize(preprocess(rawResource), customSchema);

export {
  collection,
  manifest,
  canvas,
  annotation,
  annotationPage,
  annotationCollection,
  range,
  contentResource,
  choice,
  canvasReference,
  service,
  partOf,
  annotationBody,
  rangeItem,
  RESOURCE_TYPE_MAP,
  resource,
  preprocess,
  normalizeResource as normalize,
};
