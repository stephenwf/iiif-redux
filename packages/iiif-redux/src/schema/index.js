import * as presentation2 from './presentation2';
import * as presentation3 from './presentation3';

export { presentation2, presentation3 };

// Default is 2 until 3 is ratified?
export default presentation2;
