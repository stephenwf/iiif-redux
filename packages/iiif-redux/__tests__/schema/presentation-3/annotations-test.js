import { normalize, resource } from '../../../src/schema/presentation3';

describe('Presentation 3 annotations', () => {
  test('Annotation can be normalized', () => {
    const { result, entities } = normalize(
      {
        id: 'http://example.org/iiif/annotation/1.json',
        type: 'Annotation',
        body: {
          type: 'TextualBody',
          format: 'text/plain',
          value: 'health',
          generator: 'https://omeka.dlcs-ida.org/s/ida/montague/open',
          purpose: 'tagging',
        },
      },
      resource
    );

    expect(result).toEqual({
      id: 'http://example.org/iiif/annotation/1.json',
      schema: 'annotation',
    });
    expect(entities).toMatchSnapshot();
  });
  test('Annotation with specific resource', () => {
    const { result, entities } = normalize({
      id: 'http://example.org/iiif/annotation/2.json',
      type: 'Annotation',
      body: {
        type: 'SpecificResource',
        format: 'application/html',
        generator:
          'https://omeka.dlcs-ida.org/s/ida/montague/open/topicmanager/',
        purpose: 'tagging',
        source: 'https://omeka.dlcs-ida.org/s/ida/page/topics/theme/health',
      },
    });

    expect(result).toEqual({
      id: 'http://example.org/iiif/annotation/2.json',
      schema: 'annotation',
    });
    expect(entities).toMatchSnapshot();
  });
  test('External web resource (from the data model)', () => {
    const { result, entities } = normalize({
      '@context': 'http://www.w3.org/ns/anno.jsonld',
      id: 'http://example.org/anno2',
      type: 'Annotation',
      body: {
        id: 'http://example.org/analysis1.mp3',
        format: 'audio/mpeg',
        language: 'fr',
      },
      target: {
        id: 'http://example.gov/patent1.pdf',
        format: 'application/pdf',
        language: ['en', 'ar'],
        textDirection: 'ltr',
        processingLanguage: 'en',
      },
    });

    expect(result).toEqual({
      id: 'http://example.org/anno2',
      schema: 'annotation',
    });
    expect(entities).toMatchSnapshot();
  });
});
