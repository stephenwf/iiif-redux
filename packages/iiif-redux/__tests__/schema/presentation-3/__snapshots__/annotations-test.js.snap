// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Presentation 3 annotations Annotation can be normalized 1`] = `
Object {
  "annotations": Object {
    "http://example.org/iiif/annotation/1.json": Object {
      "body": Array [
        Object {
          "id": "https://12ba91ae41c9dc7d04bae3ec44e230ec5a1d7d63",
          "schema": "contentResource",
        },
      ],
      "id": "http://example.org/iiif/annotation/1.json",
      "type": "Annotation",
    },
  },
  "contentResources": Object {
    "https://12ba91ae41c9dc7d04bae3ec44e230ec5a1d7d63": Object {
      "format": "text/plain",
      "generator": "https://omeka.dlcs-ida.org/s/ida/montague/open",
      "id": "https://12ba91ae41c9dc7d04bae3ec44e230ec5a1d7d63",
      "purpose": "tagging",
      "type": "TextualBody",
      "value": "health",
    },
  },
}
`;

exports[`Presentation 3 annotations Annotation with specific resource 1`] = `
Object {
  "annotations": Object {
    "http://example.org/iiif/annotation/2.json": Object {
      "body": Array [
        undefined,
      ],
      "id": "http://example.org/iiif/annotation/2.json",
      "type": "Annotation",
    },
  },
  "contentResources": Object {
    "undefined": Object {
      "format": "application/html",
      "generator": "https://omeka.dlcs-ida.org/s/ida/montague/open/topicmanager/",
      "purpose": "tagging",
      "source": "https://omeka.dlcs-ida.org/s/ida/page/topics/theme/health",
      "type": "SpecificResource",
    },
  },
}
`;

exports[`Presentation 3 annotations External web resource (from the data model) 1`] = `
Object {
  "annotations": Object {
    "http://example.org/anno2": Object {
      "@context": "http://www.w3.org/ns/anno.jsonld",
      "body": Array [
        Object {
          "id": "http://example.org/analysis1.mp3",
          "schema": "contentResource",
        },
      ],
      "id": "http://example.org/anno2",
      "target": Array [
        "http://example.gov/patent1.pdf",
      ],
      "type": "Annotation",
    },
  },
  "contentResources": Object {
    "http://example.org/analysis1.mp3": Object {
      "format": "audio/mpeg",
      "id": "http://example.org/analysis1.mp3",
      "language": "fr",
    },
  },
  "selectors": Object {
    "http://example.gov/patent1.pdf": "http://example.gov/patent1.pdf",
  },
}
`;
